slapo/model_schedule/api.py

```diff
@@ -0,0 +1,85 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+from .registry import get_schedule_method
+from ..schedule import create_schedule
+from ..logger import get_logger
+
+
+def apply_schedule(
+    model,
+    model_config,
+    **sch_config,
+):
+    try:
+        model_name = model_config._name_or_path
+    except Exception:
+        model_name = model_config.get("_name_or_path", None)
+    if model_name is None:
+        raise ValueError(
+            "Model name is not specified in model_config. Please use `_name_or_path` to specify the model name."
+        )
+    short_name = model_name.split("/")[-1].split("-")[0]
+    short_name = "gpt_neo" if short_name == "gpt" else short_name
+    logger = get_logger(f"{model_name}")
+    logger.info(f"Scheduling {model_name}", ranks=0)
+    logger.info(f"Schdeule config: {sch_config}", ranks=0)
+
+    # Change data type.
+    fp16 = sch_config.get("fp16", False)
+    bf16 = sch_config.get("bf16", False)
+    if fp16 and bf16:
+        raise ValueError("Cannot use both fp16 and bf16")
+    if fp16:
+        logger.info("Change model dtype to fp16", ranks=0)
+        model.half()
+    elif bf16:
+        logger.info("Change model dtype to bf16", ranks=0)
+        model.bfloat16()
+    else:
+        logger.info("Use fp32 as default model dtype", ranks=0)
+
+    group = sch_config.get("group", None)
+    sch = create_schedule(model, group=group)
+    logger.info(f"Scheduling {model_name} with TP={sch.world_size}", ranks=0)
+
+    # Tensor parallelism.
+    shard_params_fn = get_schedule_method(short_name, "shard_parameters")
```

Comment:
I didn't see the part of replacing kernels such as attention? Are they now a part of sharing? It seems not right if so.